smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/AddHttpApiKeyAuthPlugin.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.typescript.codegen.integration;
+
+import java.nio.file.Paths;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import software.amazon.smithy.codegen.core.Symbol;
+import software.amazon.smithy.codegen.core.SymbolProvider;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.knowledge.TopDownIndex;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.model.traits.HttpApiKeyAuthTrait;
+import software.amazon.smithy.model.traits.OptionalAuthTrait;
+import software.amazon.smithy.typescript.codegen.CodegenUtils;
+import software.amazon.smithy.typescript.codegen.TypeScriptSettings;
+import software.amazon.smithy.typescript.codegen.TypeScriptWriter;
+import software.amazon.smithy.utils.IoUtils;
+import software.amazon.smithy.utils.ListUtils;
+import software.amazon.smithy.utils.SmithyInternalApi;
+
+/**
+ * Add config and middleware to support a service with the @httpApiKeyAuth trait.
+ */
+@SmithyInternalApi
+public final class AddHttpApiKeyAuthPlugin implements TypeScriptIntegration {
```

Comment:
This doesn't look like it utilizes [the `@auth` trait](https://awslabs.github.io/smithy/1.0/spec/core/auth-traits.html#smithy-api-auth-trait) at all, but seems like it should. Note that [the `ServiceIndex`](https://awslabs.github.io/smithy/javadoc/1.15.0/software/amazon/smithy/model/knowledge/ServiceIndex.html) can help reason about the `@auth` trait in a clear way, notably [`getEffectiveAuthSchemes​(ToShapeId service, ToShapeId operation)`](https://awslabs.github.io/smithy/javadoc/1.15.0/software/amazon/smithy/model/knowledge/ServiceIndex.html#getEffectiveAuthSchemes(software.amazon.smithy.model.shapes.ToShapeId,software.amazon.smithy.model.shapes.ToShapeId)).

This will require additional tests for different effective schemes as well.

Reply:
Thanks @kstich . I based this class on [AddAwsAuthPlugin.java](https://github.com/aws/aws-sdk-js-v3/blob/main/codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddAwsAuthPlugin.java#L249) which also doesn't reference the `@auth` trait. I'll have to do some reading to understand how all this is meant to tie together, any further pointers you could provide would be greatly appreciated.

Reply:
@kstich based on my reading what you're suggesting seems pretty straightforward, but confirmation would be awesome.

Based on your comment, I should be using `getEffectiveAuthSchemes(service)` or `getEffectiveAuthSchemes(service, operation)` and then checking the values of the returned map for the `HttpApiKeyAuthTrait`. Yes?

It doesn't look like `getEffectiveAuthSchemes` takes `@optionalAuth` into account, so I'm guessing I need to check for that myself, at least for now. I can write some tests to validate this but want to confirm the intended behaviour rather than what the code happens to implement today.

Reply:
I've made an attempt at this, hopefully it's what you intended. The tests pass.

Reply:
Pretty close on this application, left a couple more comments 👍🏻